dropwizard-hibernate/src/main/java/io/dropwizard/hibernate/UnitOfWorkAspect.java

```diff
@@ -47,120 +55,189 @@ public UnitOfWorkAspect(Map<String, SessionFactory> sessionFactories) {
+
+            if (!errors.isEmpty()) {
+                throw errors.iterator().next().get();
             }
         } finally {
-            session = null;
-            ManagedSessionContext.unbind(sessionFactory);
+            sessions = new HashMap<>();
+            selectedSessionFactories.values().forEach(ManagedSessionContext::unbind);
+        }
+    }
+
+    private static Optional<RuntimeException> closeSession(Session session) {
+        if (session != null) {
+            try {
+                session.close();
+            } catch (HibernateException e) {
+                return Optional.of(e);
+            }
         }
+        return Optional.empty();
     }
 
-    protected void configureSession() {
-        checkNotNull(unitOfWork);
-        checkNotNull(session);
-        session.setDefaultReadOnly(unitOfWork.readOnly());
-        session.setCacheMode(unitOfWork.cacheMode());
-        session.setHibernateFlushMode(unitOfWork.flushMode());
+    protected void configureSessions() {
+        sessions.values().stream().forEach(session -> {
+            checkNotNull(unitOfWork);
+            checkNotNull(session);
+            session.setDefaultReadOnly(unitOfWork.readOnly());
+            session.setCacheMode(unitOfWork.cacheMode());
+            session.setHibernateFlushMode(unitOfWork.flushMode());
+        });
     }
 
-    private void beginTransaction(UnitOfWork unitOfWork, Session session) {
+    private void beginTransactions(UnitOfWork unitOfWork, Map<String, Session> sessions) {
         if (!unitOfWork.transactional()) {
             return;
         }
-        session.beginTransaction();
+        sessions.values().forEach(Session::beginTransaction);
     }
 
-    private void rollbackTransaction(UnitOfWork unitOfWork, Session session) {
+    private static Optional<RuntimeException> rollbackTransaction(Session session) {
+        try {
+            final Transaction txn = session.getTransaction();
+            if (txn != null && txn.getStatus().canRollback()) {
+                txn.rollback();
+            }
+            return Optional.empty();
+        } catch (RuntimeException e) {
+            return Optional.of(e);
+        }
+    }
+
+    private void rollbackTransactions(UnitOfWork unitOfWork, Map<String, Session> sessions) {
         if (!unitOfWork.transactional()) {
             return;
         }
-        final Transaction txn = session.getTransaction();
-        if (txn != null && txn.getStatus().canRollback()) {
-            txn.rollback();
+
+        //We need to ensure that the changes in all the sessions have been rollbacked
+        //and also we need to inform about the exceptions
+        List<Optional<RuntimeException>> errors = sessions.values().stream()
+            .map(UnitOfWorkAspect::rollbackTransaction)
+            .filter(Optional::isPresent)
+            .collect(Collectors.toList());
+
+        if (!errors.isEmpty()) {
+            throw errors.iterator().next().get();
```

Comment:
does this just throw the first error in the list?

Reply:
This is a tricky situation. We tried to mimic the previous behavior so If an exception is thrown during the process we should let the exception bubble up. The thing with multiple sessions is that for certain operations such as `rollback`or `session.close` we need to ensure that the operation is executed for all the sessions before letting the exception bubbles up. If we want to change this behavior in the future we are prepared to notify multiple exceptions.

Reply:
Would it make sense to throw an exception that contains references to all of the errors, if there is more than one, so no information is lost when propagating?